File: sdk/python/kcp/crypto.py
"""
KCP Cryptographic Operations

Ed25519 signing and verification + SHA-256 content hashing.
"""

import hashlib
import json
from typing import Tuple


def generate_keypair() -> Tuple[bytes, bytes]:
    """
    Generate an Ed25519 keypair.

    Returns:
        (private_key, public_key) as bytes
    """
    try:
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
        private_key = Ed25519PrivateKey.generate()
        private_bytes = private_key.private_bytes_raw()
        public_bytes = private_key.public_key().public_bytes_raw()
        return private_bytes, public_bytes
    except ImportError:
        raise ImportError(
            "KCP crypto requires 'cryptography' package. "
            "Install with: pip install cryptography"
        )


def sign_artifact(artifact_dict: dict, private_key: bytes) -> str:
    """
    Sign a KCP artifact using Ed25519.

    Args:
        artifact_dict: The artifact payload (without signature)
        private_key: Ed25519 private key bytes

    Returns:
        Hex-encoded Ed25519 signature
    """
    try:
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
    except ImportError:
        raise ImportError(
            "KCP crypto requires 'cryptography' package. "
            "Install with: pip install cryptography"
        )

    # Remove signature field if present
    payload = {k: v for k, v in artifact_dict.items() if k != "signature"}

    # Canonical JSON (sorted keys, compact)
    canonical = json.dumps(payload, sort_keys=True, separators=(",", ":"))

    # Sign
    key = Ed25519PrivateKey.from_private_bytes(private_key)
    signature = key.sign(canonical.encode("utf-8"))

    return signature.hex()


def verify_artifact(artifact_dict: dict, public_key: bytes) -> bool:
    """
    Verify the Ed25519 signature of a KCP artifact.

    Args:
        artifact_dict: The full artifact payload (with signature)
        public_key: Ed25519 public key bytes

    Returns:
        True if signature is valid, False otherwise
    """
    try:
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
        from cryptography.exceptions import InvalidSignature
    except ImportError:
        raise ImportError(
            "KCP crypto requires 'cryptography' package. "
            "Install with: pip install cryptography"
        )

    signature_hex = artifact_dict.get("signature", "")
    if not signature_hex:
        return False

    # Reconstruct canonical payload
    payload = {k: v for k, v in artifact_dict.items() if k != "signature"}
    canonical = json.dumps(payload, sort_keys=True, separators=(",", ":"))

    try:
        key = Ed25519PublicKey.from_public_bytes(public_key)
        key.verify(bytes.fromhex(signature_hex), canonical.encode("utf-8"))
        return True
    except (InvalidSignature, ValueError):
        return False


def hash_content(content: bytes) -> str:
    """
    Compute SHA-256 hash of content.

    Args:
        content: Raw content bytes

    Returns:
        Lowercase hex-encoded SHA-256 hash
    """
    return hashlib.sha256(content).hexdigest()

File: sdk/typescript/src/crypto.ts
/**
 * KCP Cryptographic Operations
 * Ed25519 signing/verification + SHA-256 hashing
 * Uses @noble/ed25519 (pure JS, no native deps) + Node.js built-in crypto
 */

import { createHash, randomBytes } from 'crypto';
import * as ed from '@noble/ed25519';
import { sha512 } from '@noble/hashes/sha512';

// noble/ed25519 v2 requires sha512 sync implementation
ed.etc.sha512Sync = (...m) => sha512(ed.etc.concatBytes(...m));

// ─── Types ───────────────────────────────────────────────────

export interface KeyPair {
  privateKey: Uint8Array; // 32 bytes
  publicKey: Uint8Array;  // 32 bytes
}

// ─── Key Generation ──────────────────────────────────────────

/**
 * Generate an Ed25519 keypair.
 * Returns 32-byte private key and 32-byte public key.
 */
export function generateKeypair(): KeyPair {
  const privateKey = ed.utils.randomPrivateKey();
  const publicKey = ed.getPublicKey(privateKey);
  return { privateKey, publicKey };
}

/**
 * Derive public key from private key.
 */
export function getPublicKey(privateKey: Uint8Array): Uint8Array {
  return ed.getPublicKey(privateKey);
}

// ─── Hashing ─────────────────────────────────────────────────

/**
 * Compute SHA-256 hash of content.
 * Returns lowercase hex string (64 chars).
 */
export function hashContent(content: Uint8Array | string): string {
  const data = typeof content === 'string' ? Buffer.from(content, 'utf8') : content;
  return createHash('sha256').update(data).digest('hex');
}

// ─── Signing ─────────────────────────────────────────────────

/**
 * Sign a KCP artifact payload using Ed25519.
 *
 * Produces canonical JSON (sorted keys, compact) of the payload
 * without the signature field, then signs it.
 *
 * Returns hex-encoded 64-byte signature (128 hex chars).
 */
export function signArtifact(
  artifactDict: Record<string, unknown>,
  privateKey: Uint8Array,
): string {
  const payload = { ...artifactDict };
  delete payload['signature'];

  // Canonical JSON: sorted keys, compact
  const canonical = canonicalJson(payload);
  const message = Buffer.from(canonical, 'utf8');

  const signature = ed.sign(message, privateKey);
  return Buffer.from(signature).toString('hex');
}

/**
 * Verify the Ed25519 signature of a KCP artifact.
 * Returns true if the signature is valid.
 */
export function verifyArtifact(
  artifactDict: Record<string, unknown>,
  publicKey: Uint8Array,
): boolean {
  const signatureHex = artifactDict['signature'] as string | undefined;
  if (!signatureHex || signatureHex.length !== 128) return false;

  const payload = { ...artifactDict };
  delete payload['signature'];

  try {
    const canonical = canonicalJson(payload);
    const message = Buffer.from(canonical, 'utf8');
    const signature = Buffer.from(signatureHex, 'hex');
    return ed.verify(signature, message, publicKey);
  } catch {
    return false;
  }
}

// ─── Helpers ─────────────────────────────────────────────────

/**
 * Produce canonical JSON: sorted keys, compact separators.
 * Mirrors Python's json.dumps(sort_keys=True, separators=(',', ':'))
 */
export function canonicalJson(obj: Record<string, unknown>): string {
  return JSON.stringify(sortObjectKeys(obj));
}

function sortObjectKeys(obj: unknown): unknown {
  if (Array.isArray(obj)) return obj.map(sortObjectKeys);
  if (obj !== null && typeof obj === 'object') {
    return Object.keys(obj as Record<string, unknown>)
      .sort()
      .reduce(
        (acc, k) => ({
          ...acc,
          [k]: sortObjectKeys((obj as Record<string, unknown>)[k]),
        }),
        {} as Record<string, unknown>,
      );
  }
  return obj;
}

/**
 * Hex-encode a Uint8Array.
 */
export function toHex(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString('hex');
}

/**
 * Decode a hex string to Uint8Array.
 */
export function fromHex(hex: string): Uint8Array {
  return new Uint8Array(Buffer.from(hex, 'hex'));
}
